### PR TITLE
HAI-2338 Implement gdpr delete dry-run for kortisto

### PR DESCRIPTION
**Pull request by corvidian**
# Description

Refactor the GDPR service dry-run to not return a list of applications. Instead, return true if the information can be deleted and throw an exception if it can't. This also means that the delete call needs to be refactored to not receive a list of applications, but fetch them from the DB again.

Besides supporting both the old and Kortisto model of GDPR service, this also has the added benefit of loading the information to be deleted in the same transaction as they are deleted in, so there's no chance of the information changing between the loading and deleting.

Implement the canDelete method for `KortistoGdprService`. This checks that the user is not on any active applications, and is not the only admin user (user with KAIKKI_OIKEUDET privileges) in the hanke. If the user shouldn't be deleted, all the conflicting applications are gathered and thrown in an exception. The controller returns the errors in the correct format to Profiili.

## Type of change

- [ ] Bug fix 
- [X] New feature 
- [ ] Other

# Instructions for testing
Can be tested with the profile-gdpr-api-tester.

# Checklist:

- [X] I have written new tests (if applicable)
- [X] I have ran the tests myself (if applicable)
- [X] I have made necessary changes to the documentation, link to confluence
 or other location: https://helsinkisolutionoffice.atlassian.net/wiki/spaces/HAI/pages/8075476993/GDPR+API#Kortiston-my%C3%B6t%C3%A4

# Other relevant info
The `deleteInfo` method will be implemented in the next PR.

### CODE REVIEW

File: services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/gdpr/KortistoGdprServiceITest.kt [hunk 1]
```diff
@@ -200,20 +207,122 @@ class KortistoGdprServiceITest(
...
+            val activeHakemus =
+                hakemusFactory.builder(USERID, hanke).inHandling().saveWithYhteystiedot {
+                    rakennuttaja()
+                }
+
+            val failure = assertFailure { gdprService.canDelete(USERID) }
+
+            failure.given { e ->
+                val messages = (e as DeleteForbiddenException).errors.map { it.message.fi }
+                assertThat(messages).single().contains(activeHakemus.applicationIdentifier!!)
```
Comment (umeetiusbaar) [inline services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/gdpr/KortistoGdprServiceITest.kt:303]: Maybe this test should assert that there is something different in the result compared to test `throws exception when there's an active application`? Or is there any difference between the results in these two tests?

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/gdpr/GdprService.kt [hunk 2]
```diff
@@ -80,7 +154,13 @@ class OldGdprService(private val applicationService: ApplicationService) : GdprS
     }
 
     @Transactional(readOnly = true)
-    override fun findApplicationsToDelete(userId: String): List<Application> {
+    override fun canDelete(userId: String): Boolean {
+        findApplicationsToDelete(userId)
+        // Will throw an exception if the information can't be deleted
```
Comment (umeetiusbaar) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/gdpr/GdprService.kt:159]: This comment might be better to be above the `findApplicationsToDelete` or on the same line with it.

File: services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/gdpr/KortistoGdprServiceITest.kt (hunk 1 above)
Comment (corvidian) [reply services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/gdpr/KortistoGdprServiceITest.kt:303]: The errors they return are exactly the same. So a request might fail on both for the same hakemus. The response is mapped to a set to get rid of the duplicates.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/gdpr/GdprService.kt (hunk 2 above)
Comment (corvidian) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/gdpr/GdprService.kt:158]: I moved the comment.

**Review by umeetiusbaar — APPROVED** [review]

**Pull request merged**